### PR TITLE
[TorchTidy] Add pattern to detect matrix alignment in fp16 AND reorganize benchmark structure

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82693
* #82261
* __->__ #82248

### CODE REVIEW

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38173079).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38173079).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38173079).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38173079).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38173079).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38263206).

File: torch/profiler/_pattern_matcher.py [hunk 1]
```diff
@@ -50,13 +51,28 @@ def eventTreeTraversal(self):
...
     def benchmark_summary(self, events: List[_ProfilerEvent]):
-        return ""
+        def format_time(time_ns: int):
+            unit_lst = ["ns", "us", "ms", "s"]
+            for unit in unit_lst:
+                if time_ns < 1000:
+                    return f"{time_ns:.2f} {unit}"
+                time_ns //= 1000
+
+        assert hasattr(self, 'benchmark'), 'Please implement benchmark()'
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:67]: Add a base `benchmark` method that raises `NotImplementedError`?

File: torch/profiler/_pattern_matcher.py [hunk 2]
```diff
@@ -471,6 +469,68 @@ def match(self, event: _ProfilerEvent):
...
+    def skip(self):
+        return not self.prof.with_stack or not self.prof.record_shapes
+
+    def match(self, event: _ProfilerEvent):
+
+        def mutiple_of(shapes, multiple):
+            return all(dim % multiple == 0 for shape in shapes
+                       for dim in shape[-2:])
+
+        if event.name() != "aten::mm" and event.name() != "aten::bmm":
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:489]:
```
if event.name() not in ("aten::mm", "aten::bmm", "aten::addmm"):
```

File: torch/profiler/_pattern_matcher.py [hunk 3]
```diff
@@ -517,20 +577,22 @@ def eventTreeBFS(event_tree: List[_ProfilerEvent]):
             stack.append(child_event)
 
 
-def report_all_anti_patterns(prof, should_benchmark: bool = False):
+def report_all_anti_patterns(prof, should_benchmark: bool = False, json_report_dir: str = None):
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:580]: `json_report_dir` Looks like this is for the next one.

**Review by robieta — APPROVED** [review]
A couple minor nits and this really should have been two PRs, but overall looks good.

File: torch/profiler/_pattern_matcher.py (hunk 1 above)
Comment (davidchencsl) [reply torch/profiler/_pattern_matcher.py:67]: I need to check if we have implemented benchmark in summary(), I haven't find a better way to solve this

File: torch/profiler/_pattern_matcher.py [hunk 4]
```diff
@@ -103,13 +101,35 @@ def go_up_until(self, event: _ProfilerEvent, predicate):
             event = event.parent
         return event
 
+    def benchmark_summary(self, events: List[_ProfilerEvent]):
+
+        def format_time(time_ns: int):
```
[the remark this answers is not in the document]
Comment (davidchencsl) [reply torch/profiler/_pattern_matcher.py]: Nice catch ^^

**Pull request closed without merging**